### ISSUE TITLE
Integrate MONAI/nnUNet model inference into `sct_deepseg`

### ISSUE THREAD

**Issue by mguaypaq** [spinalcordtoolbox/spinalcordtoolbox#4182]
This is a feature request that came out of the SCT dev meeting on 2023-07-27 ([minutes](https://docs.google.com/document/d/1ItApJQfajO2lRzOU2yenWbeRg6alfsdut3J4AVVdo78/edit#heading=h.3jazj3wbpm6z)).

We should figure out how to integrate with MONAI ([link](https://monai.io/core.html)) for deepseg models. Currently we integrate with [ivadomed](https://github.com/ivadomed/ivadomed), but the long-term intention is to migrate to MONAI instead (see meeting minutes for 2023-02-07 and subsequent meetings [here](https://docs.google.com/document/d/19mEHKOZKFdprb3UPQYxmfO46mn0bPQfOgE-WPdeF7AQ/edit#heading=h.mh06cc7ad3ul)).

Action items:
- Need an update of the `sct_deepseg` API so that it can be run under MONAI (in addition to the current ivadomed-based workflow).
- Need to coordinate with the contrast-agnostic project (@rohanbanerjee, @naga-karthik)
- Also need to discuss packaging nnUNet ([publication](https://www.nature.com/articles/s41592-020-01008-z#code-availability), [code repo](https://github.com/MIC-DKFZ/nnUNet)) – used as a part of the fMRI EPI SC segmentation project.

Questions:
- Do we import MONAI, or does MONAI import us? Or do we just need to have a compatible API?

**Comment by naga-karthik** [comment]
Thanks for opening this issue! I can briefly comment about packaging aspects of MONAI and nnUNet

> Do we import MONAI, or does MONAI import us? 

We import MONAI. I don't know if need an API or something. I created an [inference script](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/nk/monai/monai/run_inference.py) that uses a model trained using MONAI. Note, that "training with MONAI" only means that we're mainly using its augmentations and some other functions. The model at the end is indeed a PyTorch model (just like ivadomed's). This is also the reason why I said we might not need an API.

Some specifics: 
1. [These lines](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/nk/monai/monai/run_inference.py#L10-L13) are where I import what I need from MONAI. 
2. [This line](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/nk/monai/monai/run_inference.py#L80) loads a model checkpoint (this could also be a `.pt` file from PyTorch). 

> Also need to discuss packaging nnUNet

I also did try this inside a [script](https://github.com/ivadomed/model_seg_sci/tree/nk/nnunet-package/packaging) that expects list(s) of images or a dataset of images and runs inference using an nnUNet model. Also no need for an API here. The main requirement is `pip install nnunetv2` and the rest are minor packages. 

My scripts are still inside branches, I am working on finalizing them and ultimately push it to [ivadomed/utilities](https://github.com/ivadomed/utilities) repo so that others can also use it!

**Comment by mguaypaq** [comment]
Thanks for the context, the links to your inference scripts, and the highlights! So, it sounds like in both cases (MONAI and nnUNet), the main things we need to check are:
- Whether we can `pip install` suitable versions of the dependencies inside the SCT environment (in effect, merging your `requirements.txt` files for your scripts with the existing one in SCT).
- How big these extra dependencies are.

**Comment by jcohenadad** [comment]
> * merging your `requirements.txt` files for your scripts with the existing one in SCT

if we go that route, we need to "cleanup" the [requirement files](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/nk/monai/monai/requirements.txt), because presently it looks more like an output of a `pip freeze`

**Comment by naga-karthik** [comment]
Yes, I do recall the actual requirements are light. It just that most of the packages in there are automatically installed by the big packages like MONAI or nnunet. I will soon push an updated `requirements.txt` for the inference script.

**Comment by naga-karthik** [comment]
Turns out that the requirements for nnunet inference are very light! The updated `requirements.txt` can be found [here](https://github.com/ivadomed/model_seg_sci/blob/nk/nnunet-package/packaging/requirements.txt)!

**Comment by joshuacwnewton** [comment]
Hrmm... `torch=2.0.1` might be a bit of a concern, given that `ivadomed` [pins `torch>=1.8.1,<=1.11.0`](https://github.com/ivadomed/ivadomed/blob/e101ebea632683d67deab3c50dd6b372207de2a9/requirements.txt#L24).

---

On the nnUNet side of things, they [specifically pin `torch>=2.0.0`](https://github.com/MIC-DKFZ/nnUNet/blob/5af388417c36b566ea029cc8b4e057d75a8ed422/pyproject.toml#L33):

- The `torch>=2.0.0` requirement was introduced 5 months ago in https://github.com/MIC-DKFZ/nnUNet/commit/ab0b65696edb103f98b3492340e7c1e67efcc8be.
- Looking at the latest stable release ([`v2.0`](https://github.com/MIC-DKFZ/nnUNet/releases/tag/v2.0)), it falls under "torch>1.12" timeline, sadly. 
- The most recent stable version that matches ivadomed's requirements appears to be [`v1.17`](https://github.com/MIC-DKFZ/nnUNet/releases/tag/v1.7.1), which pins [`torch>=1.10`](https://github.com/MIC-DKFZ/nnUNet/blob/v1.7.1/setup.py#L12).

On the ivadomed side of things, it looks like the upper bound was added in https://github.com/ivadomed/ivadomed/pull/1137:

- There is a discussion in [this thread](https://github.com/ivadomed/ivadomed/pull/1137#discussion_r938281300) about the upper bounds.
- In that thread, there is a lot of theory-crafting about rollings vs. pinned releases, presumably under the assumption that ivadomed would continue to be actively maintained.
- But I don't actually see a strong reason proposed for 1.11.0 as the upper bound in particular. It looks like that upper bound was chosen purely because it was the minimum version necessary to support Python 3.10 (https://github.com/ivadomed/ivadomed/pull/1137/commits/24fe492e0d296f30c19c2beeba2a47505281d4d0)

So, as far as resolving this conflict goes, I presume it will be easiest on the ivadomed side of things?

**Comment by joshuacwnewton** [comment]
> So, as far as resolving this conflict goes, I presume it will be easiest on the ivadomed side of things?

Small update: Using `torch==2.0.0` passes all of the tests on both the `ivadomed` and `SCT` side of things. So, I think having MONAI/nnUNet alongside `ivadomed` should be a non-issue.

We just need to perform some maintenance tasks on the `ivadomed` side of things, then release a new patch update so that the stable release of `ivadomed` is compatible.

**Comment by naga-karthik** [comment]
Thanks for confirming this, @joshuacwnewton ! Just a thought, can we **not** pin the torch version (like `torch=2.0.0`) and keep just `torch`? I know that version pinning is always recommended but considering my script for running inference, it uses the most basic functions of torch which must be there in every version. What do you think of this?

**Comment by joshuacwnewton** [comment]
Oh, my apologies! The version pinning wasn't intended to be the final fix -- I just wanted to make sure that `2.0.0` would be chosen in my testing. (If I didn't pin to 2.0.0, then `pip` might fall back to a different version, masking some failures during my tests.)

I absolutely agree that we should leave the upper bound unpinned.

**Comment by mguaypaq** [comment]
This morning @naga-karthik and I did a quick and dirty test run of installing the model inference dependencies and scripts into SCT, and running an example inference. TL;DR: it worked! Here are some notes.

We did:
1. Clone SCT.
2. Manually create a conda environment for it, with `python==3.9` and `libstdcxx-ng`, `libffi`.
3. Get ready to use pip: `pip install --upgrade pip setuptools wheel`
4. Clone `ivadomed/ivadomed`, checkout the branch `jn/test-torch-compatibility_base-kk-chore-imageio`.
5. Get ivadomed and its dependencies: `pip install -e ivadomed`
   - Note: this installed `torch==2.0.0`, which was later overwritten with `torch==2.0.0+cpu`.
6. Clone `sct-pipeline/contrast-agnostic-softseg-spinalcord`, checkout the branch `nk/monai`.
7. Copy a few files into SCT:
   - `monai/run_inference_single_image.py` to `spinalcordtoolbox/scripts/run_inference_single_image.py` (edited to import `spinalcordtoolbox.monai.models` instead of `models`)
   - `monai/models.py` to `spinalcordtoolbox/monai/models.py` (edited to import `spinalcordtoolbox.monai.building_blocks` instead of `building_blocks`)
   - `monai/building_blocks.py` to `spinalcordtoolbox/monai/building_blocks.py`
8. Edit `sct/requirements.txt`:
   - Comment out the `ivadomed~=2.9.6` pin.
   - Add:
     ```
     dynamic_network_architectures==0.2
     joblib==1.3.0
     loguru==0.7.0
     monai[nibabel]==1.2.0
     scipy==1.11.2
     --extra-index-url https://download.pytorch.org/whl/cpu
     torch==2.0.0+cpu
     ```
     Note that `numpy==1.24.4` is **not** in this list, since SCT pins `numpy<1.24`.
9. `pip install -r sct/requirements.txt`
10. `pip install -e sct`
11. Copy scripts to `sct/bin` and download data to `sct/data`.
12. All is green: `sct_check_dependencies`
13. Put the following file in `sct/data/agnostic/`:
    ```
    duke/temp/muena/contrast-agnostic/final_monai_model/nnunet_nf=32_DS=1_opt=adam_lr=0.001_AdapW_CCrop_bs=2_64x192x320_20230918-2253/best_model_loss.ckpt
    ```
14. Get an image from `sct_example_data` and run inference:
    ```bash
    python "${SCT_DIR}/spinalcordtoolbox/scripts/run_inference_single_image.py" --path-img t2.nii.gz --chkp-path "${SCT_DIR}/data/agnostic/" --path-out .
    ```
    This generated a couple of warnings, but basically worked, and produced a fine-looking inference (visually inspected in fsleyes):
    ```
    $SCT_DIR/lib/python3.9/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'libc10_cuda.so: cannot open shared object file: No such file or directory'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
      warn(
    2023-09-28 12:12:04.197 | WARNING  | __main__:main:146 - GPU not available, using CPU instead
    2023-09-28 12:12:04.201 | INFO     | __main__:main:159 - Saving results to: .
    2023-09-28 12:12:12.643 | INFO     | __main__:main:226 - Saving subject: t2
    $SCT_DIR/python/sct/lib/python3.9/site-packages/monai/utils/deprecate_utils.py:321: FutureWarning: monai.transforms.io.array SaveImage.__init__:resample: Current default value of argument `resample=True` has been deprecated since version 1.1. It will be changed to `resample=False` in version 1.3.
      warn_deprecated(argname, msg, warning_category)
    2023-09-28 12:12:12.769 | INFO     | __main__:main:256 - ========================================================
    2023-09-28 12:12:12.770 | INFO     | __main__:main:257 -       Inference Time per Subject: 6.90s
    2023-09-28 12:12:12.770 | INFO     | __main__:main:258 - ========================================================
    ```

**Comment by jcohenadad** [comment]
A-M-A-Z-I-N-G!  Bravo team! 👏

I'm wondering though: why installing ivadomed? Do we still depend on it?

**Comment by joshuacwnewton** [comment]
> I'm wondering though: why installing ivadomed? Do we still depend on it?

As the codebase currently stands, `sct_deepseg` is essentially just a wrapper for `ivadomed`'s inference functionality. So, unless we plan to remove support for all of deepseg's current models, we will need to continue to maintain `ivadomed` (or at least the inference subset of `ivadomed`).

**Comment by jcohenadad** [comment]
> As the codebase currently stands, sct_deepseg is essentially just a wrapper for ivadomed's inference functionality. So, unless we plan to remove support for all of deepseg's current models, we will need to continue to maintain ivadomed (or at least the inference subset of ivadomed).

Right. In my mind we would eventually replace _almost_ all current models, but this is something that will be done in the longer term. I might be able to recruit a student for doing only that.

**Comment by naga-karthik** [comment]
Just so everyone is on the same page - the next steps after [this](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4182#issuecomment-1739680083) experiment include: 

* [ ] Cleaning the [inference script](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/nk/monai/monai/run_inference_single_image.py) currently on my branch`nk/monai` on the `sct-pipeline/contrast-agnostic-softseg-spinalcord` repo to make it a standalone `.py` script (right now it imports the model class from another file)
* [ ] Merging my branch to `main`
* [ ] Creating a release with the current model (though is not the _final_ model yet, still have to run it on more seeds)

**Comment by joshuacwnewton** [comment]
I'm currently working on testing the inference scripts for the MONAI and nnUnet projects to make sure they're compatible with `ivadomed` in the same virtual environment (see https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4231), and I had a quick question: What is the difference between the `monai`-based nnUnet architecure and the `nnunetv2` package?

For context: There are two main inference scripts across the softseg/rootlets/SCI projects: 

1. [`run_inference_single_image.py`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4231/files#diff-99c4a6ed40afb0d3c2b54059503030c3e413d3b60aa3bfbb7c26a1f636dc3508) (MONAI - softseg)
2. [`run_inference_single_subject.py`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4231/files#diff-8e5c9655ecd0bfa9e802cf29ba0882eabc9650a4a7013e3fcf5dbe64a44d3961) (nnUNet - rootlets/SCI)
 
Given the similar names of the scripts, I was curious if I could try to combine them to create a unified interface. Both scripts seem to use the nnUNet architecture, however the implementations are quite different!

So, I was just curious about how we got to this state. Is there a reason we couldn't use a similar structure (either the `monai` package or the `nnunetv2` package) for all of these projects? (For example, maybe the two groups of projects evolved differently, and different packages were available at different times?)

**Comment by naga-karthik** [comment]
> What is the difference between the monai-based nnUnet architecure and the nnunetv2 package?

In short, monai-based nnunet is [just the network architecture](https://github.com/MIC-DKFZ/dynamic-network-architectures/blob/main/dynamic_network_architectures/architectures/unet.py#L15) without the bells-and-whistles involved when installing the whole `nnunet` package

> I was curious if I could try to combine them to create a unified interface.

Hmm, I would not recommend that. The monai-based inference (bullet 1) just gets the predictions through the nnunet model (without even depending on the `nnunet` package) but keeps the data loading, preprocessing, saving etc. in-house. Whereas, the nnuent-based inference (bullet 2) lets nnunet handles everything from loading to saving. The problem with the latter is it is very bulky (i.e. inference takes from 3-5 mins to upto 20 mins on a CPU) depends on the arguments you provide. There are several issues on the nnunet repo about inference being too slow and there are some hacky ways to get around it (i.e. disable test-time augmentation, decrease sliding-window overlap, etc.). These are not good solutions so I wanted to move away from that and try to do everything with monai-based inference.

> So, I was just curious about how we got to this state.

IMO the beauty of nnunet is its out-of-the-box solution: curate dataset --> pip install nnunet --> train model --> done. This conveniently skips the hard parts of machine learning where you spend days/months try to come up with a good model architecture/hyperparameters. But, that's the end of the story. It's only recently that we realized that nnunet is heavily optimized for training and segmentation performance but not for production/deployment. Also, nnunet source code is kind of hard to tinker with. This is why are starting to do everything in monai, where we have control over each step of the inference pipeline.

**Comment by joshuacwnewton** [comment]
Thank you so much for the explanation!

> These are not good solutions so I wanted to move away from that and try to do everything in-house with monai-based inference.

Seeing as we want to move away from the `nnunetv2` package and prioritize `monai`, do we have a plan for the existing `nnunetv2`-based projects (SCI, rootlets)? 

For example, I notice MONAI has a [`monai.apps.nnunet.NNUnetV2Runner`](https://docs.monai.io/en/stable/apps.html#module-monai.apps.nnunet.__main__) class with a [`predict`](https://docs.monai.io/en/stable/_modules/monai/apps/nnunet/nnunetv2_runner.html#nnUNetV2Runner.predict) method. Though, it seems to use the exact same `nnunetv2.inference.predict_from_raw_data` module as the inference script linked above (bullet 2). So, even if we used MONAI's nnUNetV2 interface (along with the existing trained model folders), we would still be installing the `nnunetv2` package with the slow/bulky inference, right? :thinking:

**Comment by naga-karthik** [comment]
> do we have a plan for the existing nnunetv2-based projects (SCI, rootlets)?

Depends on the type of project I believe and whether we expect a lot of people using it. For example, for [SCI segmentation](https://github.com/ivadomed/model_seg_sci), our collaborators already tried the nnunet-based inference and did not come back with any issues/complaints. Hence, we decided to keep it this way for the SCI model at least. 

There is also a bigger question of how much time one can spend on porting from nnuent to monai (depends on how 'urgent' the project is)

> MONAI's nnUNetV2 interface (along with the existing trained model folders), we would still be installing the nnunetv2 package with the slow/bulky inference, right?

Yeah, I looked at this too but sadly came to the same (correct) conclusion as you have. MONAI's app for nnunetv2 just provides a wrapper around nnunet with some config-file-based training functionalities. The nnunet still installation remains.

**Comment by joshuacwnewton** [comment]
> There is also a bigger question of how much time one can spend on porting from nnuent to monai (depends on how 'urgent' the project is)

This is something we can discuss in the Thursday SCT meeting agenda, I suppose?

But, we can always start with an nnunet-based implementation in SCT for just those two projects, and leave it as a future TODO to migrate them from `nnunetv2` to MONAI if need be?

**Comment by naga-karthik** [comment]
Yes, that sounds good @joshuacwnewton! This can be added to this week's meeting agenda

**Comment by joshuacwnewton** [comment]
Marking this issue as `user requested` so that I will know to return to it when following up on past forum posts as part of the 6.2 release procedure. (i.e. https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4301)
